pyro/infer/predictive.py

```diff
@@ -66,15 +67,15 @@ def _predictive(model, posterior_samples, num_samples, return_sites=None,
     for site in model_trace.stochastic_nodes + model_trace.observation_nodes:
         append_ndim = max_plate_nesting - len(model_trace.nodes[site]["fn"].batch_shape)
         site_shape = (num_samples,) + (1,) * append_ndim + model_trace.nodes[site]['value'].shape
-        if isinstance(return_sites, (list, tuple, set)):
+        if return_sites is not None:
             if site in return_sites:
                 return_site_shapes[site] = site_shape
         else:
-            if (return_sites is not None) or (site not in reshaped_samples):
-                return_site_shapes[site] = site_shape
+            # include all sites by default
```

Comment:
@neerajprad I meant this case.

Reply:
That's a mistake, thanks for catching! Can I just use the `if site not in posterior_samples` check here, or do we really need the sentinel `''` value?

Reply:
I think using `''` is not needed if we raise an exception where both `guide` and `posterior_samples` are specified. I don't know in which situation users need both of them. :D

Reply:
I think you original cases are indeed needed. :) Seems like we would need to distinguish between three cases, but I'm just using a `()` instead of `''`.
 - when `return_sites` is non empty we return samples from it.
 - when `return_sites = ()` (default), we return samples not in posterior samples.
 - when `return_sites = None` and a `guide` argument is provided (only used internally when a guide is provided), we return all sites. This is important because a guide will pass in `posterior_samples` but we should return those samples to the user.

Reply:
Yeah, that looks good to me!